Port to Vue.js

Axolotle opened the pull request:
This is a complete rewrite of the web-admin with vue.js.
Only the files containing the translations have been moved because the nature of the port prevented any reuse of the code (switch to ES6 syntax + removal of jquery + specificity of vue.js).
So I just created a new directory `app/` in which you can find the new files.

## CHANGELOG

- less time waiting for server's responses thanks to a cache system for users, domains, groups and permissions
- skeletons while waiting for a view's initial api queries
- new console/history that displays messages by actions with some information and is sizable at drag
- improved app catalog search (all filters are accessible at the same time)
- more client-side validation on forms
- ability to search in views that are lists (users, domains, services, etc.)
- new error pages for actions and route access (the first as a modal, the other as a view)
- web-admin settings:
    - primary language selector
    - secondary language selector
    - cache toggling
    - page transition animation toggling
- uniformization of the look:
    - should be fully responsive
    - add an icon on every cards
    - improved some views with new buttons/inputs
    - uniformize sizing, alignment and stuff
    - reworked the post install look
    - added buttons to toggle cards that are collapsable
    - card infos with highlighter
- misc:
    - internationalized dates/durations
    - ability to search in permissions and users in "Groups and permissions" when adding a permission/user
    - ability to create multiple mail aliases and mail forwards on one form submit.
    - ability to relog after reboot without reloading the page (asked to when the server is accessible)
    - manage labels and access url forms are now directly into the app info view
    - changed the font to "Fira" and "Fira Code"
    - add translations to the footer menu items
    - can't remove a main domain with message
    - app catalog also search in descriptions

for dev:
- new dev server with hot-reload (live update at code modification without full rebuild)
- validation helpers
- generic but customisable components for:
    - basic views (skeleton and automatic fetching)
    - search views (views that display lists that can be filtered)
    - views top bar with kindof automatic css handling depending on the content
    - card
    - card headers with slots to insert buttons (with kindof automatic css handling)
    - form
    - form items (inputs, checkboxes, selects)
- automatic modal handling for user confirmation (with promises)
- dev mode toggling in settings to display routes to experimental features

## PREVIEW
Actual preview page tree (will be updated if i'm not lazy)
![yunohost pages](https://autre.space/ressources/ynh/pageScreen-low.png)

## PR Status
Needs test

### TEST NEEDED

- migrations (skipid, disclaimer and all)
- postinstall
- errors
- ConfigPanel

### OPINION NEEDED

- are you ok with skeletons or should i revert to a spinner or something ?
- websocket messages display during action:
    - when you trigger an action the history bar is automatically opened and displays only the last message
    - when the action is done it close itself after a short delay
    - you can still completely open the history to see all messages during an action
- do we trigger the update automatically when clicking on "System update" ?
- do we trigger the diagnosis automatically

### TODO/FIX

- [x] integrate to ynh-dev env (see https://github.com/YunoHost/ynh-dev/pull/60)
- [x] update debian install (not tested, see https://github.com/YunoHost/yunohost-admin/pull/316/commits/ef54685e2f7521c3139a6d98a98ac46d6bd5a9fc)
- [ ] update `travis.yml` ?
- [ ] error on app install checkbox (not converted to 0:1 if not touched)
- [ ] error "APIBadRequestError: There is already a YunoHost operation running. Please wait for it to finish before running another one." not displayed since it is a code 400 (errors 400 have some special behavior and displayed in forms but this error can be catch on no form views.
- [ ] add a message to explain why the "I don't have a domain name… " is disabled if user has already a dyndns domain
- [ ] remove the old app
- [ ] much test

## How to test

You can test the code within the lxc or from outside.

*From outside the lxc*
```bash
# In yunohost-admin/app/
# add a .env file with your serveur ip:
echo "VUE_APP_IP='Y.O.U.R.I.P'" > .env

# install dependencies
npm install

# run webpack-dev-server with hot-reloads and moulte stuff
npm run serve
```

*From within the lxc with https://github.com/YunoHost/ynh-dev/pull/60 changes*
```bash
# from `ynh-dev/`
git clone -b enh-vuejs --single-branch https://github.com/Axolotle/yunohost-admin yunohost-admin-vue

# from the lxc
./ynh-dev use-git yunohost-admin-vue
```

Axolotle reviewed (commented):
https://github.com/YunoHost/yunohost-admin/pull/316/commits/ef54685e2f7521c3139a6d98a98ac46d6bd5a9fc don't know if that's ok !?

alexAubin commented:
>ef54685 don't know if that's ok !?

There's also some epic black dirty magic here that's gonna need to be adapted but we can see this later ;P

https://github.com/YunoHost/vinaigrette/blob/master/scripts/pbuilder/hooks/D80patch-yunohost-admin#L16-L58

Axolotle commented:
I've updated some translations key changes to all locales. And added some new ones in `en.json`, sometimes with nested keys.
Most of the translations real changes in `en.json` are the replacement of `%s` by named variables like `{name}` or ordered variables like `{0}` since it is way more readable for translators and allow changing the arguments order. 

I don't know how to reflect these changes in the others translation files, should we try to do it ourself (but some translations file aren't up to date we're likely to make mistakes) or should we let translators do it ?

alexAubin commented:
Hey there !

Here's a first round of feedback ! Overall : this is pretty impressive considering it's a full rework, yet everything is pretty much working, super smooth and with very few regressions. There are many great improvements such as the new modal while an operation is ongoing (especially for app installs with the progress bar, this is smooth af), the search bar for users / domain / ..., the rework of the app catalog into a single page with all the filters at the top, and many things I have yet to fully realize (I did not even open the code so far, and if i remember correctly many things happened behind the scenes too). (Also toggle switches are awesome)

Of course it's easier to make a list of "what's to improve / fix" but hopefully that's not too depressing ¯\\\_\(ツ)_/¯

----

### General

- (already known) Locale string full of %s ... We certainly don't want to ask translators to go through all the translated strings just for this (this will take a while and would annoy everybody) Maybe I can try to work on a small python script that magically identifies the `{foo}` from en.json to replace the corresponding `%s` on translations, i already worked on that kind of stuff in the past.

- The pacman goes way too fast which is stressful..., shouldn't it be contained inside the "Waiting for server response" modal instead of the entire screen ?

- UX: The "wait for response" modal disappears too quickly ... sometimes there are important information or warning shown, and it's just not possible to read it because it goes away too fast. I understand that the message goes in the "History" tab, but it's not 100% obvious and still pretty annoying to see a message yet not being able to read it right away. *Oooor* you may not realize that the operation produced a warning (for example: when you add a custom per-user permission, it may say something about the permission already being provided through another group - the warning itself is debatable, but it remains that with the current system, everybody will completely miss this warning).

- ... But then I don't know what else to suggest for this :/ Overall I don't know what to do or think about the whole question of "how do we properly display logs / info regarding what's going on (or previous errors/warnings) to the users" which turns out
   to be absolutely not a trivial design problematic.

- History bar: having to click right on the "caret" icon is not ideal if you want to open the history tab regularly - ideally would be nice to be able to  just click on the bar and have it show up 

- History bar: ideally one should be able to resize the height :s

- History bar: it's a bit weird that the way messages are displayed (with a colored icon on the right) is different than how it's done in the modal (with a colored background). Not 100% troublesome but a bit weird when you think about it. Though (c.f. later) I agree that the colorful background is just too much.

- Cosmetics, user/app/service lists : I believe it would improve readability to have the username or appname (basically the main info) as bold (and/or the secondary informations as grey) ? 

- Cosmetics : I find the colors of the default bootstrap theme(?) are too flashy, in particular "warnings" and "primary" (and in particular i always find it weird that the font for warning is black compared to white for others). I find it's also the case for background colors which are overwhelming, especially on the diagnosis screen, where I would tend to tweak the CSS to have lighter backgrounds...

- Cosmetics : in the various info lists/panels, that thing about dashes between the key and values feels a bit weird / cumbersome but maybe that's just me ... but that also makes me want to polish some stuff because sometimes the same info is displayed 2 or 3 times and that's feels a bit drafty ...

- UX : the button "Share with Yunopaste" absolutely needs to be in green to make it visible to users (on the support, we have to tell people to share the damn log all the time - so really the mental sanity is at stake here...). It also needs to be green to create a sense of positivity about sharing the log :/. 

### Permission view

- UX : the '-' sign to remove the permission only appearing when you hover the mouse sounds like bad UX ... non-technical people tend to have trouble discovering that kind of mechanism, and can you really use this on mobile ?

- [ux] When thinking about "how do I add a new permission", the "search" icon is not so obvious that this how you *add* a new
  permission ... (also when there's no permission that you can add, it feels like there's a bug because nothing happens when you click on it)

### User creation view

- [ux] the help texts when definining the email address and passwords are now only in black/grey, while before they were in info/warning alerts. Imho it's important that these texts / explanations are highlighted in some way to really encourage the user to read them :s

### User edit view

- When trying to add aliases or forwards, the "Save" button is always disabled and there's no way to save (in fact, even when editing anything else in the form)

### App list view

- Ideally, we could make it so that the id is not displayed if it's the same than the label ? Otherwise it's pretty much always repeated and it's weird. (This is not a regression of this PR, but before there was a bigger distinction between the label and the id so it was less obvious)

### App catalog

- [cosmetics] In the app cards, the border between the buttons is  black which looks weird compared to the external border which is lightgrey ?

- [cosmetics/ux] The "not maintained" flag takes too much place imho, and so far we tend to try to keep it discreet to not scare people about that flag.

- [cosmetics] I would also keep the app names in bold ?

- [ux] : the search is great, but also quite slow when you're typing text in the input. (My computer has a reasonably powerful CPU/RAM). Ideally would be nice if there's some optimization that can be done

- The button "install" is displayed on multiple lines for notworkings apps (because of the small warning icon)

- Some apps have a *green* "not yet working" badge ? (show that for : Invidious, chtikynotes, encryptic ... sounds like apps flagged as "working" but with an undefined level, imho those should be treated as notworking or low quality until a level is defined)

### App install

- [ux] I need to think more about how the app install failures is handled and what to do with this ... (this is also related to the whole discussion about the modal vs. history). There's at least two cases for this : 
      - 1. install parameters provided are refused, in which case you want to stay on the form and ask the user to fix it
      - 2. the effective install failed, and you'd like to redirect directly to the log view (I was hoping to work on this at some point, the server could send a reference to the log in its response, which could then be picked up by the webadmin to craft a redirect)

### App info view

- [cosmetics] the info about currently allowed users + button to go manage the permission is oddly placed ... maybe we could try to find a way to integrate it like the other info in the same panel, with a button at the end of the line ?

- [cosmetics] there's a small help text "Manage permissions labels in the portal." which in fact doesn't bring any more info than what's already written so i would simply remove it

### Service view

- for some services (e.g. ssh) the button 'stop' is disable but should *not even be displayed*. Imho having the button displayed is more frustrating than anything else, + it somewhat allows people to easily bypass the restriction by changing the css property ?

- when trying to stop/start service, the pacman/modal hangs indefinitely despite the view being updated in the background

### Log view

- [cosmetics] the format for displaying the info is not consistent with other places of the webdmin, here the key/values are on different lines ?

- [ux] having the 'Show more lines' button in blue is imho not something we want (for non-technical users, we want their attention to be focused on sharing the logs rather than looking at the logs in details ... except for 'obvious' messages at the end like lack of space or unsupported arch...) I would keep something as neutral as possible like white/grey ?

- [ux] having *two* buttons for sharing the logs (one in the panel title, the other at the bottom of the logs) was a feature and not a bug, because people are like to miss it if there's only one, and we *reaaaaally* want to people to find this easily to preserve our mental sanity.

### Others / misc

- Change admin password view : Apparently there a small debug-related variable dump remaining

- Webadmin settings:
   - "Locale" sounds more like a technical word, why not use "Language" ?
   - I'm not sure we really need a mechanism for "fallback locale" ? (It doesn't sound that much important, and one would have to explain what this means ...)

- Backup create : It feels weird that the "System" and "Applications" titles are in green ?

- Footer centered on logging screen, but not after ? Is it a bug or a feature ?

- When trying to login, if wrong password is entered, the pacman+modal loops forever

- At some point we were discussing using the Fira font but I see it's still using Source ?

Axolotle commented:
Thanks for the feedback !

- [ ] replace `%s` with named variables in translations
- [ ] update base colors and lighter colors
- [ ] waiting modal:
  - [ ] limit pacman translation to the modal width
  - [ ] fix automatic closing of the modal (in discussion)
- [ ] history:
  - [ ] show on bar click
  - [ ] ability to resize the history elem height
  - [ ] uniformize ws message level color with waiting modal
- [ ] user/app/service lists : main info as bold, secondary informations as grey
- [ ] display the best guess for an app name (label/id) in AppList
- [ ] AppCatalog:
  - [ ] uniformize card border color
  - [ ] reduce size of `Not maintained`
  - [ ] app name in bold ?
  - [ ] optimize app search computation
  - [ ] fix warning icon in `Install` button overflow
  - [ ] fix green app color for some `not yet working` apps (ex: Invidious, chtikynotes, encryptic, …) => not working or low quality
- [ ] AppInstall:
  - [ ] manage users/permission button as an item in the descriptive list
  - [ ] remove useless `Manage permissions labels in the portal.` label
- [ ] `Share with yunopaste` in green
- [ ] rework group/permission view
  - [ ] rework selectize
  - [ ] find a way to make it accessible
  - [ ] do not hide the `-` button
- [ ] fix UserEdit aliases/forward addition
- [ ] add an optional argument to FormField to display a description with a text/background color
- [ ] 
- [ ] 
- [ ] 


Making a list out of the obvious things to fix:
- 
- 
    - 
    - 
    - 
- 
- 
- ServiceInfo:
    - remove `stop` buttons instead of disabling it
    - fix infinite waiting modal when stopping/running a service
- LogInfo:
    - uniformize descriptive list
    - `show more lines` button in gray
    - return to double `Share with Yunopast` button (in top and bottom of the view)
- remove debug log in ToolAdminpw
- rename `Locale` to `Language`
- do not use green in BackupCreate `Application` + `System`
- fix wrong password in login infinite waiting modal
- add Fira


> ### General
>
>     * UX: The "wait for response" modal disappears too quickly ... sometimes there are important information or warning shown, and it's just not possible to read it because it goes away too fast. I understand that the message goes in the "History" tab, but it's not 100% obvious and still pretty annoying to see a message yet not being able to read it right away. _Oooor_ you may not realize that the operation produced a warning (for example: when you add a custom per-user permission, it may say something about the permission already being provided through another group - the warning itself is debatable, but it remains that with the current system, everybody will completely miss this warning).
>
>     * ... But then I don't know what else to suggest for this :/ Overall I don't know what to do or think about the whole question of "how do we properly display logs / info regarding what's going on (or previous errors/warnings) to the users" which turns out
>       to be absolutely not a trivial design problematic.

Yep, for now the modal stays open and an `Ok` button appears when the api returns an error that is not a code 400. The problem is that there's no difference between an error 400 produced by a wrong form value (that you want to show directly in the form and not in the modal) and something that went wrong in an install process for example (that you want to show in the modal or redirect to the error log as you proposed).  
It could be great if the API could return some more information about the error (at least if its related to the user input or an internal error and at best the name of the falsy form argument so the form can display this error directly under the falsy input)

It could be something like that:
- automatically close the modal if 400 is wrong user input and show the error in the form
- let the modal opened if "internal" 400 or redirect to the log
- let the modal opened if other error code or websocket warnings/errors

Also websocket warning that are targeted at packagers should'nt be shown, right? (could be a options in the webadmin settings)

>     * Cosmetics : in the various info lists/panels, that thing about dashes between the key and values feels a bit weird / cumbersome but maybe that's just me ... but that also makes me want to polish some stuff because sometimes the same info is displayed 2 or 3 times and that's feels a bit drafty ...

Yep, i was trying to find a way to link items in those descriptive lists, i can remove the dashes since the gray backgrounds should be enough.

> ### Permission view
>
>     * UX : the '-' sign to remove the permission only appearing when you hover the mouse sounds like bad UX ... non-technical people tend to have trouble discovering that kind of mechanism, and can you really use this on mobile ?
>
>     * [ux] When thinking about "how do I add a new permission", the "search" icon is not so obvious that this how you _add_ a new
>       permission ... (also when there's no permission that you can add, it feels like there's a bug because nothing happens when you click on it)

This view really needs some rework indeed. The minus is also accessible thru focus (tab) but yeah, it won't work for mobile.  
This view is not accessible (and it was already the case), I do not know yet how to make it work with screen reader (i'll try with controls and regions) and the custom selectize component i made (combo of input and select) needs to be improved.


> ### User creation view
>
>     * [ux] the help texts when definining the email address and passwords are now only in black/grey, while before they were in info/warning alerts. Imho it's important that these texts / explanations are highlighted in some way to really encourage the user to read them :s

For now the "description" under an input can't be displayed with a color but i can modify this component. I moved those inside the fieldset since screen readers weren't able to read those descriptions at all in the Sammy version.

> ### App install
>
>     * [ux] I need to think more about how the app install failures is handled and what to do with this ... (this is also related to the whole discussion about the modal vs. history). There's at least two cases for this :
>       - 1. install parameters provided are refused, in which case you want to stay on the form and ask the user to fix it
>       - 2. the effective install failed, and you'd like to redirect directly to the log view (I was hoping to work on this at some point, the server could send a reference to the log in its response, which could then be picked up by the webadmin to craft a redirect)

Yep, same as above proposition.

> ### Others / misc
>     * Webadmin settings:
>       
>       * "Locale" sounds more like a technical word, why not use "Language" ?
>       * I'm not sure we really need a mechanism for "fallback locale" ? (It doesn't sound that much important, and one would have to explain what this means ...)

I will change `Locale` for `Language`, but i think we should keep the fallback picker (with an help text if needed). Users may use a regional language, and may not speak English but another language: Someone may use `Català` as a primary language, and prefer to fallback to `Français` or `Español` depending on its geographical location, and not to `English` (which is the hidden last fallback locale if none of the first two had the translation). This can also be the case for multiple countries that has multiple official languages that are not `English`.

>     * Footer centered on logging screen, but not after ? Is it a bug or a feature ?

Kinda feature ? Since there's no version info, it is centered like in small screen size.

>     * At some point we were discussing using the Fira font but I see it's still using Source ?

Yep, i didn't add it yet. But there's no font at all actually, seems like you have `Source` installed on your system (and the css for the font still reference its name)

zamentur commented:
>     * [ ]  fix UserEdit aliases/forward addition => WONTFIX, No bug for me ?

I found the same issue. In fact you need to enter 0 in the quota mailbox to get the save button available.

Axolotle commented:
> > ```
> > * [ ]  fix UserEdit aliases/forward addition => WONTFIX, No bug for me ?
> > ```
> 
> I found the same issue. In fact you need to enter 0 in the quota mailbox to get the save button available.

Yep, i had some problems with the validation that forced me to ask for a number (no empty value). I can investigate more if that's a problem.

kay0u commented on app/src/views/tool/ToolLog.vue:
```diff
@@ -0,0 +1,136 @@
+      <b-button
+        v-if="moreLogsAvailable"
+        variant="white" class="w-100 rounded-0"
+        @click="$ref.view.fetchQueries()"
```
```suggestion
        @click="$refs.view.fetchQueries()"
```

I have no idea what I'm doing, but it works better like this

kay0u commented on app/src/views/tool/ToolLog.vue:
```diff
@@ -0,0 +1,136 @@
+      this.description = log.description
+
+      const levels = ['ERROR', 'WARNING', 'SUCCESS', 'INFO']
+      this.logs = log.logs.map(line => {
+        for (const level of levels) {
+          if (line.includes(level + ' -')) {
+            return `<span class="alert-${level === 'ERROR'
+              ? 'danger'
+              : level.toLowerCase()}">${line}</span>`
+          }
+        }
+        return line
```
```suggestion
      function escapeHtml (unsafe) {
        return unsafe
              .replace(/&/g, '&amp;')
              .replace(/</g, '&lt;')
              .replace(/>/g, '&gt;')
              .replace(/"/g, '&quot;')
              .replace(/'/g, '&#039;')
      }
      this.logs = log.logs.map(line => {
        for (const level of levels) {
          if (line.includes(level + ' -')) {
            return escapeHtml(`<span class="alert-${level === 'ERROR'
              ? 'danger'
              : level.toLowerCase()}">${line}</span>`)
          }
        }
        return escapeHtml(line)
```

Escape html in logs (is it the right way to do it?)

Axolotle replied:
```suggestion
      this.logs = log.logs.map(line => {
        const escaped = escapeHtml(line)
        for (const level of levels) {
          if (line.includes(level + ' -')) {
            return `<span class="alert-${level === 'ERROR'
              ? 'danger'
              : level.toLowerCase()}">${escaped}</span>`
          }
        }
        return escaped
```

We can escape the line indeed in case the logs contains HTML, but escaping the `<span class="alert-">` that allows highlighting will break it !

Also there is multiple potential problems (that existed in the Sammy version) :
- Some errors sent by the api contains HTML (sent as text)
- Some translations values contains HTML
Those are not escaped because it may contains formatting and links and comes from locales files, if the translations are carefully reviewed before merging it may be ok but [it is not recommended](https://kazupon.github.io/vue-i18n/guide/formatting.html#html-formatting). vue-i18n recommends to use [component interpolation](https://kazupon.github.io/vue-i18n/guide/interpolation.html) but this breaks sentences into multiple chunks so translators may not know what to do with it.

alexAubin commented:
![ezgif-5-a2085414a9](https://user-images.githubusercontent.com/4533074/105096582-9887b180-5aa7-11eb-87d3-35fbe7ac280a.gif)
